binaries/download_binaries.py

```diff
@@ -0,0 +1,152 @@
+the platforms we build for and output a yaml file for platform<->binaries mapping.
+"""
+
+import time
+import os
+import yaml
+from typing import List
+from threading import Thread
+
+import streamer_binaries
+import requests
+
+
+YAML_OUTPUT_FILE = 'package_data.yaml'
+
+
+def download_github_release_binaries(
+    output_dir: str,
+    github_api_end_points: List[str],
+    only_download_files_starting_with: List[str] = ['']):
+  """Downloads the latest releases for a github repo.
+  `output_dir` is the directory that files will be downloaded in.
+  This function is blocking, it will wait until all the assets are downloaded.
+  
+  `github_api_end_point` should be a releases end point for a github repo.
+
+  Check the Releases API at https://docs.github.com/en/rest/reference/repos#releases
+
+  `only_download_files_starting_with` is an optional argument, if passed, the function
+  will try to only download files starting with one of this list elements.
+  """
+
+  def download_binary(download_url: str, downloaded_files: List[str]):
+    """Downloads a file and write it to the file system."""
+
+    file_name = download_url.split('/')[-1]
+    # Don't try to download files other than the ones starting with
+    # one of the strings in `only_download_files_starting_with`.
+    if not file_name.startswith(tuple(only_download_files_starting_with)):
+      return
+    file_path = os.path.join(output_dir, file_name)
+    # Download with `stream=True` so not to use so much memory.
+    with requests.get(url=download_url, stream=True) as res:
+      # Was the connection successful?
+      try:
+        res.raise_for_status()
+      except Exception as ex:
+        raise RuntimeError('\'{}\' couldn\'t be downloaded\nException: {}\n'.format(
+            download_url, ex))
+      print('downloading', file_name)
+      with open(file_path, 'wb') as binary:
+        for content in res.iter_content(chunk_size=65536):
+          binary.write(content)
+      print(file_name, 'is downloaded.')
+    # Set executable permissions for the downloaded binaries.
+    default_permissions = 0o755
+    os.chmod(file_path, default_permissions)
+    # Add the this file we just downloaded to the downloaded files.
+    downloaded_files.append(file_name)
+
+  assets: List[dict] = []
+  for end_point in github_api_end_points:
+    res = requests.get(end_point,
+                       # This header is recommended by github docs.
+                       headers={'Accept': 'application/vnd.github.v3+json'})
+    res.raise_for_status()
+    res_in_json = res.json()
+    # `res_in_json[0]` will grab the latest release.
+    assets.extend(res_in_json[0]['assets'])
+
+  download_threads: List[Thread] = []
+  downloaded_files: List[str] = []
+
+  # Loop over all the assets and download them.
+  for asset in assets:
+    download_url: str = asset['browser_download_url']
+    download_threads.append(Thread(target=download_binary,
```

Comment:
It seems like overkill to use threads to download binaries.  I would prefer something simpler, even if it's somewhat slower.